File: boss_bridge/test_client/proxy_server.js
/**
 * 简单的CORS代理服务器 (Node.js版本)
 * 用于解决测试客户端的跨域问题
 * 
 * 使用方法:
 *   node proxy_server.js
 * 
 * 然后在测试客户端中，将服务器地址改为:
 *   http://localhost:8888/proxy
 */

const http = require('http');
const https = require('https');
const url = require('url');

// 默认目标服务器地址
const DEFAULT_TARGET = process.env.RADIUS_SERVER_URL || 'http://10.11.6.92:8090/services/IServiceUopBossToTvManager';

const PORT = process.argv[2] ? parseInt(process.argv[2]) : 8888;

const server = http.createServer((req, res) => {
    // 处理CORS预检请求
    if (req.method === 'OPTIONS') {
        res.writeHead(200, {
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
            'Access-Control-Allow-Headers': 'Content-Type, SOAPAction',
            'Access-Control-Max-Age': '3600'
        });
        res.end();
        return;
    }

    // 只处理 /proxy 路径
    if (!req.url.startsWith('/proxy')) {
        res.writeHead(404, { 'Content-Type': 'text/plain' });
        res.end('Not Found');
        return;
    }

    // 获取目标URL
    let targetUrl = DEFAULT_TARGET;
    const parsedUrl = url.parse(req.url, true);
    
    if (parsedUrl.query.url) {
        targetUrl = parsedUrl.query.url;
    } else if (req.headers['x-target-url']) {
        targetUrl = req.headers['x-target-url'];
    }

    console.log(`[${new Date().toISOString()}] ${req.method} ${req.url} -> ${targetUrl}`);

    // 解析目标URL
    const target = url.parse(targetUrl);
    const isHttps = target.protocol === 'https:';
    const httpModule = isHttps ? https : http;

    // 收集请求体
    let body = [];
    req.on('data', chunk => {
        body.push(chunk);
    });

    req.on('end', () => {
        const requestBody = Buffer.concat(body);

        // 创建转发请求的选项
        const options = {
            hostname: target.hostname,
            port: target.port || (isHttps ? 443 : 80),
            path: target.path,
            method: req.method,
            headers: {}
        };

        // 复制原始请求头（排除一些不需要的）
        for (const key in req.headers) {
            if (key.toLowerCase() !== 'host' && 
                key.toLowerCase() !== 'connection' &&
                key.toLowerCase() !== 'content-length') {
                options.headers[key] = req.headers[key];
            }
        }

        // 如果有请求体，设置Content-Length
        if (requestBody.length > 0) {
            options.headers['Content-Length'] = requestBody.length;
        }

        // 发送请求
        const proxyReq = httpModule.request(options, (proxyRes) => {
            // 设置CORS响应头
            res.writeHead(proxyRes.statusCode, {
                'Access-Control-Allow-Origin': '*',
                'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
                'Access-Control-Allow-Headers': 'Content-Type, SOAPAction',
                'Content-Type': proxyRes.headers['content-type'] || 'text/xml; charset=utf-8'
            });

            // 转发响应体
            proxyRes.pipe(res);
        });

        proxyReq.on('error', (err) => {
            console.error(`代理错误: ${err.message}`);
            res.writeHead(500, {
                'Access-Control-Allow-Origin': '*',
                'Content-Type': 'text/plain'
            });
            res.end(`代理错误: ${err.message}`);
        });

        // 发送请求体
        if (requestBody.length > 0) {
            proxyReq.write(requestBody);
        }
        proxyReq.end();
    });

    req.on('error', (err) => {
        console.error(`请求错误: ${err.message}`);
        res.writeHead(500, {
            'Access-Control-Allow-Origin': '*',
            'Content-Type': 'text/plain'
        });
        res.end(`请求错误: ${err.message}`);
    });
});

server.listen(PORT, () => {
    console.log('代理服务器已启动');
    console.log(`监听地址: http://localhost:${PORT}`);
    console.log(`代理路径: http://localhost:${PORT}/proxy`);
    console.log(`\n在测试客户端中配置:`);
    console.log(`  服务器IP: localhost`);
    console.log(`  端口: ${PORT}`);
    console.log(`  路径: /proxy`);
    console.log(`\n按 Ctrl+C 停止服务器`);
});

File: boss_bridge/test_client/proxy_server.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
简单的CORS代理服务器
用于解决测试客户端的跨域问题

使用方法:
    python proxy_server.py

然后在测试客户端中，将服务器地址改为:
    http://localhost:8888/proxy
"""

from http.server import HTTPServer, BaseHTTPRequestHandler
import urllib.request
import urllib.parse
import json

class ProxyHandler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        """处理CORS预检请求"""
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type, SOAPAction')
        self.send_header('Access-Control-Max-Age', '3600')
        self.end_headers()

    def do_GET(self):
        """处理GET请求（用于测试连接）"""
        if self.path.startswith('/proxy'):
            # 从查询参数获取目标URL
            query = urllib.parse.urlparse(self.path).query
            params = urllib.parse.parse_qs(query)
            target_url = params.get('url', [None])[0]
            
            if not target_url:
                self.send_error(400, "Missing 'url' parameter")
                return
            
            try:
                # 转发请求
                req = urllib.request.Request(target_url)
                with urllib.request.urlopen(req, timeout=10) as response:
                    content = response.read()
                    
                    # 返回响应，添加CORS头
                    self.send_response(200)
                    self.send_header('Access-Control-Allow-Origin', '*')
                    self.send_header('Content-Type', response.headers.get('Content-Type', 'text/xml'))
                    self.end_headers()
                    self.wfile.write(content)
            except Exception as e:
                self.send_error(500, str(e))
        else:
            self.send_error(404)

    def do_POST(self):
        """处理POST请求（用于SOAP请求）"""
        if self.path == '/proxy' or self.path.startswith('/proxy?'):
            # 读取请求体
            content_length = int(self.headers.get('Content-Length', 0))
            body = self.rfile.read(content_length)
            
            # 从查询参数或请求头获取目标URL
            target_url = None
            if '?' in self.path:
                query = urllib.parse.urlparse(self.path).query
                params = urllib.parse.parse_qs(query)
                target_url = params.get('url', [None])[0]
            
            # 如果没有在URL中，尝试从自定义头获取
            if not target_url:
                target_url = self.headers.get('X-Target-URL')
            
            # 如果还是没有，使用默认配置
            if not target_url:
                # 默认RADIUS服务器地址，可以从环境变量或配置文件读取
                import os
                target_url = os.environ.get('RADIUS_SERVER_URL', 
                    'http://10.11.6.92:8090/services/IServiceUopBossToTvManager')
            
            try:
                # 创建转发请求
                req = urllib.request.Request(target_url, data=body)
                
                # 复制原始请求头
                for header, value in self.headers.items():
                    if header.lower() not in ['host', 'content-length', 'connection']:
                        req.add_header(header, value)
                
                # 转发请求
                with urllib.request.urlopen(req, timeout=30) as response:
                    content = response.read()
                    
                    # 返回响应，添加CORS头
                    self.send_response(200)
                    self.send_header('Access-Control-Allow-Origin', '*')
                    self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
                    self.send_header('Access-Control-Allow-Headers', 'Content-Type, SOAPAction')
                    self.send_header('Content-Type', response.headers.get('Content-Type', 'text/xml; charset=utf-8'))
                    self.end_headers()
                    self.wfile.write(content)
            except urllib.error.HTTPError as e:
                # 转发HTTP错误
                error_content = e.read() if e.fp else b''
                self.send_response(e.code)
                self.send_header('Access-Control-Allow-Origin', '*')
                self.send_header('Content-Type', 'text/xml; charset=utf-8')
                self.end_headers()
                self.wfile.write(error_content)
            except Exception as e:
                self.send_error(500, f"Proxy error: {str(e)}")
        else:
            self.send_error(404)

    def log_message(self, format, *args):
        """自定义日志格式"""
        print(f"[{self.address_string()}] {format % args}")

def run(port=8888):
    """启动代理服务器"""
    server_address = ('', port)
    httpd = HTTPServer(server_address, ProxyHandler)
    print(f"代理服务器已启动")
    print(f"监听地址: http://localhost:{port}")
    print(f"代理路径: http://localhost:{port}/proxy")
    print(f"\n在测试客户端中配置:")
    print(f"  服务器IP: localhost")
    print(f"  端口: {port}")
    print(f"  路径: /proxy")
    print(f"\n按 Ctrl+C 停止服务器")
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        print("\n服务器已停止")
        httpd.server_close()

if __name__ == '__main__':
    import sys
    port = 8888
    if len(sys.argv) > 1:
        try:
            port = int(sys.argv[1])
        except ValueError:
            print("用法: python proxy_server.py [端口号]")
            sys.exit(1)
    run(port)
